[turborepo] `turbo prune` produces a broken pnpm-lock.yaml lockfile

### What version of Turborepo are you using?

1.7.0

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Linux

### Describe the Bug

Running `turbo prun` will produce a broken pnpm lock file under: `./out/pnpm-lock.yaml` that will be missing a child dependency if it was marked as `npm:<SOME_DEPENDENCY>`. Running `pnpm build` will produce the proper lock file `./pnpm-lock.yaml`.

This behavior is leading to this error ` ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY `

### Expected Behavior

lock file should be valid

### To Reproduce

Clone the repo and install using `pnpm`

```
git clone git@github.com:omarahm3/turbo-pnpm-missing-dependency.git
cd turbo-pnpm-missing-dependency
pnpm i
```

In this repo i used turbo repo [basic template](https://github.com/vercel/turbo/blob/main/examples/basic/README.md) and just added these packages:

- [plotly.js](https://github.com/plotly/plotly.js)
- [react-plotly.js](https://github.com/plotly/react-plotly.js/)
- as dev dependency @types/react-plotly.js

now run:

```
pnpm turbo prune --scope=web
cd out
pnpm i --frozen-lockfile
```

you'll get this error:

```
Scope: all 5 workspace projects
Lockfile is up to date, resolution step is skipped
 ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  Broken lockfile: no entry for '/@plotly/regl/2.1.2' in pnpm-lock.yaml

This issue is probably caused by a badly resolved merge conflict.
To fix the lockfile, run 'pnpm install --no-frozen-lockfile'.
```

Now i'm not really sure if this has something to do with this issue, but looking at the problematic dependency's - plotly.js - package.json file especially [this line](https://github.com/plotly/plotly.js/blob/master/package.json#L109):

```
    "regl": "npm:@plotly/regl@^2.1.2",
```

it seems that this format `npm:<PACKAGE>` is what makes it get ignored by `turbo prune` command

---

This also might be related to this #2118 but im not really sure.

### Reproduction Repo

https://github.com/omarahm3/turbo-pnpm-missing-dependency

Found out that this issue is indeed related to child package's dependency version is prefixed with `npm:` in which it doesn't get resolved by `PnpmLockfile:resolveSpecifier`.

I was able to add a case to produce the exact scenario:

**Fixture:**

```yaml
lockfileVersion: 5.4
importers:
  apps/web:
    specifiers:
      somepackage: ^2.17.1
    dependencies:
      somepackage: 2.17.1
packages:
  /somepackage/2.17.1:
    resolution: {integrity: sha512-Tgjk1FrA4TIJAMEFlJqq9swyRsdwKas90DNHoV6MZ+uxmcEO7O/kdEonu9LEIrmiNiuIQgGxwNERH7jrJxjAlQ==}
    dependencies:
      somepackage-child: /@somepackage/somepackage-child/2.1.2
    transitivePeerDependencies:
      - supports-color
    dev: false

  /@somepackage/somepackage-child/2.1.2:
    resolution: {integrity: sha512-Mdk+vUACbQvjd0m/1JJjOOafmkp/EpmHjISsopEz5Av44CBq7rPC05HHNbYGKVyNUF2zmEoBS/TT0pd0SPFFyw==}
    dev: false
```

**Test:**

```go
func Test_SpecifierResolution(t *testing.T) {
  ...
	cases := []Case{
    ...
		{workspacePath: "apps/web", pkg: "somepackage-child", specifier: "/@somepackage/somepackage-child/2.1.2", version: "/@somepackage/somepackage-child/2.1.2", found: true},
	}
  ...
}
```

Running the test now will fail:

```bash
--- FAIL: Test_SpecifierResolution (0.01s)
    pnpm_lockfile_test.go:91: assertion failed: false (actualFound bool) != true (testCase.found bool): somepackage-child@/@somepackage/somepackage-child/2.1.2
FAIL
exit status 1
FAIL    github.com/vercel/turbo/cli/internal/lockfile   0.009s
```

Digging more into the function, found that function was using `formatPnpmKey(name, specifier)` and trying to resolve `/somepackage-child//@somepackage/somepackage-child/2.1.2` which obviously does not exist. However if directly we got the package using `specifier` (`/@somepackage/somepackage-child/2.1.2`) only it will be found.

So modifying the function to this helped passing the test

```go
func (p *PnpmLockfile) resolveSpecifier(workspacePath turbopath.AnchoredUnixPath, name string, specifier string) (string, bool, error) {
	// Check if the specifier is already a resolved version
	_, ok := p.Packages[formatPnpmKey(name, specifier)]
	if ok {
		return specifier, true, nil
	}
	if _, ok := p.Packages[specifier]; ok {
		return specifier, true, nil
	}
  ...
}
```

output now:

```bash
PASS
ok      github.com/vercel/turbo/cli/internal/lockfile   0.008s
```

Now obviously this is a bad solution but i decided to go with it and see if the actual `prune --scope=web` will work, and obviously it did, but the still the same issue existed.

This is because on `PnpmLockfile:ResolvePackage` there was still a call to `formatPnpmKey(name, resolvedVersion)` and it was still trying to get the package using `(p.Packages)[key]` which is basically looking for `(p.Packages)["/somepackage-child//@somepackage/somepackage-child/2.1.2"]` and it was not found.

Now modifying the function again:

```go
func (p *PnpmLockfile) ResolvePackage(workspacePath turbopath.AnchoredUnixPath, name string, version string) (Package, error) {
	resolvedVersion, ok, err := p.resolveSpecifier(workspacePath, name, version)
	if !ok || err != nil {
		return Package{}, err
	}
	key := formatPnpmKey(name, resolvedVersion)
	if entry, ok := (p.Packages)[key]; ok {
		var version string
		if entry.Version != "" {
			version = entry.Version
		} else {
			version = resolvedVersion
		}
		return Package{Key: key, Version: version, Found: true}, nil
	}
	if entry, ok := (p.Packages)[resolvedVersion]; ok {
		var version string
		if entry.Version != "" {
			version = entry.Version
		} else {
			version = resolvedVersion
		}
		return Package{Key: version, Version: version, Found: true}, nil
	}

	return Package{}, nil
}
```

and this did it, now prune is resolving the package as it should.

This code changes is indeed not correct at all, but its just a result of a quick investigation, I hope this will help somehow understanding the issue better and hence resolving it.

--- a/cli/internal/lockfile/pnpm_lockfile.go
+++ b/cli/internal/lockfile/pnpm_lockfile.go
@@ -202,6 +202,11 @@ func DecodePnpmLockfile(contents []byte) (Lockfile, error) {
 
 // ResolvePackage Given a package and version returns the key, resolved version, and if it was found
 func (p *PnpmLockfile) ResolvePackage(workspacePath turbopath.AnchoredUnixPath, name string, version string) (Package, error) {
+	// Check if version is a key
+	if _, ok := p.Packages[version]; ok {
+		return Package{Key: version, Version: getVersionFromKey(version), Found: true}, nil
+	}
+
 	resolvedVersion, ok, err := p.resolveSpecifier(workspacePath, name, version)
 	if !ok || err != nil {
 		return Package{}, err
@@ -222,6 +227,11 @@ func (p *PnpmLockfile) ResolvePackage(workspacePath turbopath.AnchoredUnixPath, 
 
 // ResolvePackage Given a package and version returns the key, resolved version, and if it was found
 func (p *PnpmLockfileV6) ResolvePackage(workspacePath turbopath.AnchoredUnixPath, name string, version string) (Package, error) {
+	// Check if version is a key
+	if _, ok := p.Packages[version]; ok {
+		return Package{Key: version, Version: getVersionFromKeyV6(version), Found: true}, nil
+	}
+
 	resolvedVersion, ok, err := p.resolveSpecifier(workspacePath, name, version)
 	if !ok || err != nil {
 		return Package{}, err
@@ -585,4 +595,20 @@ func formatPnpmKey(name string, version string) string {
 
 func formatPnpmKeyV6(name string, version string) string {
 	return fmt.Sprintf("/%s@%s", name, version)
+}
+
+func getVersionFromKey(key string) string {
+	atIndex := strings.LastIndex(key, "/")
+	if atIndex == -1 || len(key) == atIndex+1 {
+		return ""
+	}
+	return key[atIndex+1:]
+}
+
+func getVersionFromKeyV6(key string) string {
+	atIndex := strings.LastIndex(key, "@")
+	if atIndex == -1 || len(key) == atIndex+1 {
+		return ""
+	}
+	return key[atIndex+1:]
 }

--- a/cli/internal/lockfile/pnpm_lockfile_test.go
+++ b/cli/internal/lockfile/pnpm_lockfile_test.go
@@ -216,15 +216,25 @@ func Test_PnpmPrunePatchesV6(t *testing.T) {
 }
 
 func Test_PnpmAbsoluteDependency(t *testing.T) {
-	contents, err := getFixture(t, "pnpm-absolute.yaml")
-	assert.NilError(t, err)
+	type testCase struct {
+		fixture string
+		key     string
+	}
+	testcases := []testCase{
+		{"pnpm-absolute.yaml", "/@scope/child/1.0.0"},
+		{"pnpm-absolute-v6.yaml", "/@scope/child@1.0.0"},
+	}
+	for _, tc := range testcases {
+		contents, err := getFixture(t, tc.fixture)
+		assert.NilError(t, err, tc.fixture)
 
-	lockfile, err := DecodePnpmLockfile(contents)
-	assert.NilError(t, err)
+		lockfile, err := DecodePnpmLockfile(contents)
+		assert.NilError(t, err, tc.fixture)
 
-	pkg, err := lockfile.ResolvePackage(turbopath.AnchoredUnixPath("packages/a"), "child", "/@scope/child/1.0.0")
-	assert.NilError(t, err, "resolve")
-	assert.Assert(t, pkg.Found)
-	assert.DeepEqual(t, pkg.Key, "/@scope/child/1.0.0")
-	assert.DeepEqual(t, pkg.Version, "1.0.0")
+		pkg, err := lockfile.ResolvePackage(turbopath.AnchoredUnixPath("packages/a"), "child", tc.key)
+		assert.NilError(t, err, "resolve")
+		assert.Assert(t, pkg.Found, tc.fixture)
+		assert.DeepEqual(t, pkg.Key, tc.key)
+		assert.DeepEqual(t, pkg.Version, "1.0.0")
+	}
 }

--- a/cli/internal/lockfile/pnpm_lockfile_test.go
+++ b/cli/internal/lockfile/pnpm_lockfile_test.go
@@ -214,3 +214,17 @@ func Test_PnpmPrunePatchesV6(t *testing.T) {
 
 	assert.Equal(t, len(prunedLockfile.Patches()), 1)
 }
+
+func Test_PnpmAbsoluteDependency(t *testing.T) {
+	contents, err := getFixture(t, "pnpm-absolute.yaml")
+	assert.NilError(t, err)
+
+	lockfile, err := DecodePnpmLockfile(contents)
+	assert.NilError(t, err)
+
+	pkg, err := lockfile.ResolvePackage(turbopath.AnchoredUnixPath("packages/a"), "child", "/@scope/child/1.0.0")
+	assert.NilError(t, err, "resolve")
+	assert.Assert(t, pkg.Found)
+	assert.DeepEqual(t, pkg.Key, "/@scope/child/1.0.0")
+	assert.DeepEqual(t, pkg.Version, "1.0.0")
+}

--- a/cli/internal/lockfile/testdata/pnpm-absolute-v6.yaml
+++ b/cli/internal/lockfile/testdata/pnpm-absolute-v6.yaml
@@ -1,0 +1,18 @@
+lockfileVersion: "6.0"
+importers:
+  packages/a:
+    dependencies:
+      "@scope/parent":
+        specifier: ^1.0.0
+        version: 1.0.0
+
+packages:
+  /@scope/parent@1.0.0:
+    resolution: { integrity: junk }
+    dependencies:
+      child: /@scope/child@1.0.0
+    dev: false
+
+  /@scope/child@1.0.0:
+    resolution: { integrity: junk }
+    dev: false

--- a/cli/internal/lockfile/testdata/pnpm-absolute.yaml
+++ b/cli/internal/lockfile/testdata/pnpm-absolute.yaml
@@ -1,0 +1,17 @@
+lockfileVersion: 5.4
+importers:
+  packages/a:
+    specifiers:
+      "@scope/parent": ^1.0.0
+    dependencies:
+      somepackage: 1.0.0
+packages:
+  /@scope/parent/1.0.0:
+    resolution: { integrity: junk }
+    dependencies:
+      child: /@scope/child/1.0.0
+    dev: false
+
+  /@scope/child/1.0.0:
+    resolution: { integrity: junk }
+    dev: false

